Add some resiliency to the collection of reachability metadata

**Is your feature request related to a problem? Please describe.**

We build 10s of native images on a nightly basis and [roughly one night in three](https://ge.spring.io/scans/failures?failures.failureClassification=all_failures&failures.failureMessage=Execution%20failed%20for%20task%20*%0A%3E%20Failed%20to%20query%20the%20value%20of%20task%20*%20property%20%27metadataService%27.%0A%20%20%20%3E%20Failed%20to%20create%20service%20%27nativeConfigurationService%27.%0A%20%20%20%20%20%20%3E%20Could%20not%20create%20an%20instance%20of%20type%20org.graalvm.buildtools.gradle.internal.GraalVMReachabilityMetadataService.%0A%20%20%20%20%20%20%20%20%20%3E%20java.net.ConnectException:%20Connection%20timed%20out&search.relativeStartTime=P28D&search.timeZoneId=Europe%2FLondon) (at the time of writing) we'll see a failure due to a connection timeout when downloading reachability metadata. I would like our builds to be more robust.

**Describe the solution you'd like**
Builds could be made more robust by configuring a longer connection timeout or by retrying failures that should be transient.

**Describe alternatives you've considered**
I considered some configuration on the Gradle side to retry the task that's failing. Some research suggests that this isn't something that Gradle offers, at least not without resorting to hacks. It also wouldn't benefit other users of `GraalVMReachabilityMetadataService`.

**Additional context**

<details>
<summary>Connection timeout stack trace</summary>

```
Caused by: java.net.ConnectException: Connection timed out
	at org.graalvm.buildtools.gradle.internal.GraalVMReachabilityMetadataService.newRepository(GraalVMReachabilityMetadataService.java:128)
	at org.graalvm.buildtools.gradle.internal.GraalVMReachabilityMetadataService.<init>(GraalVMReachabilityMetadataService.java:97)
	at org.graalvm.buildtools.gradle.internal.GraalVMReachabilityMetadataService$Inject.<init>(Unknown Source)
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
	at jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at org.gradle.internal.instantiation.generator.AsmBackedClassGenerator$InvokeConstructorStrategy.newInstance(AsmBackedClassGenerator.java:1913)
	at org.gradle.internal.instantiation.generator.AbstractClassGenerator$GeneratedClassImpl$GeneratedConstructorImpl.newInstance(AbstractClassGenerator.java:506)
	at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.doCreate(DependencyInjectingInstantiator.java:64)
	at org.gradle.internal.instantiation.generator.DependencyInjectingInstantiator.newInstance(DependencyInjectingInstantiator.java:55)
	at org.gradle.api.services.internal.BuildServiceProvider.instantiate(BuildServiceProvider.java:152)
	at org.gradle.api.services.internal.BuildServiceProvider.instantiate(BuildServiceProvider.java:141)
	at org.gradle.api.services.internal.BuildServiceProvider.getInstance(BuildServiceProvider.java:128)
	at org.gradle.api.services.internal.BuildServiceProvider.calculateOwnValue(BuildServiceProvider.java:121)
	at org.gradle.api.internal.provider.AbstractMinimalProvider.calculateValue(AbstractMinimalProvider.java:107)
	at org.gradle.api.internal.provider.DefaultProperty.calculateValueFrom(DefaultProperty.java:128)
	at org.gradle.api.internal.provider.DefaultProperty.calculateValueFrom(DefaultProperty.java:26)
	at org.gradle.api.internal.provider.AbstractProperty.doCalculateValue(AbstractProperty.java:133)
	at org.gradle.api.internal.provider.AbstractProperty.calculateOwnValue(AbstractProperty.java:127)
	at org.gradle.api.internal.provider.AbstractMinimalProvider.calculateOwnPresentValue(AbstractMinimalProvider.java:72)
	at org.gradle.api.internal.provider.AbstractMinimalProvider.get(AbstractMinimalProvider.java:92)
	at org.graalvm.buildtools.gradle.tasks.CollectReachabilityMetadata.copyReachabilityMetadata(CollectReachabilityMetadata.java:132)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.gradle.internal.reflect.JavaMethod.invoke(JavaMethod.java:125)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.doExecute(StandardTaskAction.java:58)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:51)
	at org.gradle.api.internal.project.taskfactory.StandardTaskAction.execute(StandardTaskAction.java:29)
	at org.gradle.api.internal.tasks.execution.TaskExecution$3.run(TaskExecution.java:236)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:29)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$1.execute(DefaultBuildOperationRunner.java:26)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.run(DefaultBuildOperationRunner.java:47)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:68)
	at org.gradle.api.internal.tasks.execution.TaskExecution.executeAction(TaskExecution.java:221)
	at org.gradle.api.internal.tasks.execution.TaskExecution.executeActions(TaskExecution.java:204)
	at org.gradle.api.internal.tasks.execution.TaskExecution.executeWithPreviousOutputFiles(TaskExecution.java:187)
	at org.gradle.api.internal.tasks.execution.TaskExecution.execute(TaskExecution.java:165)
	at org.gradle.internal.execution.steps.ExecuteStep.executeInternal(ExecuteStep.java:89)
	at org.gradle.internal.execution.steps.ExecuteStep.access$000(ExecuteStep.java:40)
	at org.gradle.internal.execution.steps.ExecuteStep$1.call(ExecuteStep.java:53)
	at org.gradle.internal.execution.steps.ExecuteStep$1.call(ExecuteStep.java:50)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:204)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:199)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:73)
	at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:50)
	at org.gradle.internal.execution.steps.ExecuteStep.execute(ExecuteStep.java:40)
	at org.gradle.internal.execution.steps.RemovePreviousOutputsStep.execute(RemovePreviousOutputsStep.java:68)
	at org.gradle.internal.execution.steps.RemovePreviousOutputsStep.execute(RemovePreviousOutputsStep.java:38)
	at org.gradle.internal.execution.steps.CancelExecutionStep.execute(CancelExecutionStep.java:41)
	at org.gradle.internal.execution.steps.TimeoutStep.executeWithoutTimeout(TimeoutStep.java:74)
	at org.gradle.internal.execution.steps.TimeoutStep.execute(TimeoutStep.java:55)
	at org.gradle.internal.execution.steps.CreateOutputsStep.execute(CreateOutputsStep.java:51)
	at org.gradle.internal.execution.steps.CreateOutputsStep.execute(CreateOutputsStep.java:29)
	at org.gradle.internal.execution.steps.CaptureStateAfterExecutionStep.executeDelegateBroadcastingChanges(CaptureStateAfterExecutionStep.java:124)
	at org.gradle.internal.execution.steps.CaptureStateAfterExecutionStep.execute(CaptureStateAfterExecutionStep.java:80)
	at org.gradle.internal.execution.steps.CaptureStateAfterExecutionStep.execute(CaptureStateAfterExecutionStep.java:58)
	at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:48)
	at org.gradle.internal.execution.steps.ResolveInputChangesStep.execute(ResolveInputChangesStep.java:36)
	at org.gradle.internal.execution.steps.BuildCacheStep.executeWithoutCache(BuildCacheStep.java:181)
	at org.gradle.internal.execution.steps.BuildCacheStep.lambda$execute$1(BuildCacheStep.java:71)
	at org.gradle.internal.Either$Right.fold(Either.java:175)
	at org.gradle.internal.execution.caching.CachingState.fold(CachingState.java:59)
	at org.gradle.internal.execution.steps.BuildCacheStep.execute(BuildCacheStep.java:69)
	at org.gradle.internal.execution.steps.BuildCacheStep.execute(BuildCacheStep.java:47)
	at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:36)
	at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:25)
	at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:36)
	at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:22)
	at org.gradle.internal.execution.steps.SkipUpToDateStep.executeBecause(SkipUpToDateStep.java:110)
	at org.gradle.internal.execution.steps.SkipUpToDateStep.lambda$execute$2(SkipUpToDateStep.java:56)
	at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:56)
	at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:38)
	at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:73)
	at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:44)
	at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:37)
	at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:27)
	at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:89)
	at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:50)
	at org.gradle.internal.execution.steps.ValidateStep.execute(ValidateStep.java:102)
	at org.gradle.internal.execution.steps.ValidateStep.execute(ValidateStep.java:57)
	at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:76)
	at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:50)
	at org.gradle.internal.execution.steps.SkipEmptyWorkStep.executeWithNoEmptySources(SkipEmptyWorkStep.java:254)
	at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:91)
	at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:56)
	at org.gradle.internal.execution.steps.RemoveUntrackedExecutionStateStep.execute(RemoveUntrackedExecutionStateStep.java:32)
	at org.gradle.internal.execution.steps.RemoveUntrackedExecutionStateStep.execute(RemoveUntrackedExecutionStateStep.java:21)
	at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsStartedStep.execute(MarkSnapshottingInputsStartedStep.java:38)
	at org.gradle.internal.execution.steps.LoadPreviousExecutionStateStep.execute(LoadPreviousExecutionStateStep.java:43)
	at org.gradle.internal.execution.steps.LoadPreviousExecutionStateStep.execute(LoadPreviousExecutionStateStep.java:31)
	at org.gradle.internal.execution.steps.AssignWorkspaceStep.lambda$execute$0(AssignWorkspaceStep.java:40)
	at org.gradle.api.internal.tasks.execution.TaskExecution$4.withWorkspace(TaskExecution.java:281)
	at org.gradle.internal.execution.steps.AssignWorkspaceStep.execute(AssignWorkspaceStep.java:40)
	at org.gradle.internal.execution.steps.AssignWorkspaceStep.execute(AssignWorkspaceStep.java:30)
	at org.gradle.internal.execution.steps.IdentityCacheStep.execute(IdentityCacheStep.java:37)
	at org.gradle.internal.execution.steps.IdentityCacheStep.execute(IdentityCacheStep.java:27)
	at org.gradle.internal.execution.steps.IdentifyStep.execute(IdentifyStep.java:44)
	at org.gradle.internal.execution.steps.IdentifyStep.execute(IdentifyStep.java:33)
	at org.gradle.internal.execution.impl.DefaultExecutionEngine$1.execute(DefaultExecutionEngine.java:76)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:139)
	at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:128)
	at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:77)
	at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
	at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:51)
	at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
	at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:57)
	at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:204)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:199)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:73)
	at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
	at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:69)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:322)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:309)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:302)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:288)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:462)
	at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:379)
	at org.gradle.execution.plan.DefaultPlanExecutor.process(DefaultPlanExecutor.java:115)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph.executeWithServices(DefaultTaskExecutionGraph.java:135)
	at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph.execute(DefaultTaskExecutionGraph.java:120)
	at org.gradle.execution.SelectedTaskExecutionAction.execute(SelectedTaskExecutionAction.java:35)
	at org.gradle.execution.DryRunBuildExecutionAction.execute(DryRunBuildExecutionAction.java:51)
	at org.gradle.execution.BuildOperationFiringBuildWorkerExecutor$ExecuteTasks.call(BuildOperationFiringBuildWorkerExecutor.java:54)
	at org.gradle.execution.BuildOperationFiringBuildWorkerExecutor$ExecuteTasks.call(BuildOperationFiringBuildWorkerExecutor.java:43)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:204)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$CallableBuildOperationWorker.execute(DefaultBuildOperationRunner.java:199)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:66)
	at org.gradle.internal.operations.DefaultBuildOperationRunner$2.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:157)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.execute(DefaultBuildOperationRunner.java:59)
	at org.gradle.internal.operations.DefaultBuildOperationRunner.call(DefaultBuildOperationRunner.java:53)
	at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:73)
	at org.gradle.execution.BuildOperationFiringBuildWorkerExecutor.execute(BuildOperationFiringBuildWorkerExecutor.java:40)
	at org.gradle.internal.build.DefaultBuildLifecycleController.lambda$executeTasks$6(DefaultBuildLifecycleController.java:167)
	at org.gradle.internal.model.StateTransitionController.doTransition(StateTransitionController.java:247)
	at org.gradle.internal.model.StateTransitionController.lambda$tryTransition$7(StateTransitionController.java:174)
	at org.gradle.internal.work.DefaultSynchronizer.withLock(DefaultSynchronizer.java:44)
	at org.gradle.internal.model.StateTransitionController.tryTransition(StateTransitionController.java:174)
	at org.gradle.internal.build.DefaultBuildLifecycleController.executeTasks(DefaultBuildLifecycleController.java:167)
	at org.gradle.internal.build.DefaultBuildWorkGraphController$DefaultBuildWorkGraph.runWork(DefaultBuildWorkGraphController.java:190)
	at org.gradle.internal.work.DefaultWorkerLeaseService.withLocks(DefaultWorkerLeaseService.java:249)
	at org.gradle.internal.work.DefaultWorkerLeaseService.runAsWorkerThread(DefaultWorkerLeaseService.java:109)
	at org.gradle.composite.internal.DefaultBuildController.doRun(DefaultBuildController.java:172)
	at org.gradle.composite.internal.DefaultBuildController.access$000(DefaultBuildController.java:47)
	at org.gradle.composite.internal.DefaultBuildController$BuildOpRunnable.run(DefaultBuildController.java:191)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:49)
```

</details>

--- a/common/utils/src/main/java/org/graalvm/buildtools/utils/ExponentialBackoff.java
+++ b/common/utils/src/main/java/org/graalvm/buildtools/utils/ExponentialBackoff.java
@@ -1,0 +1,128 @@
+/*
+ * Copyright 2003-2021 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.graalvm.buildtools.utils;
+
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+
+/**
+ * An utility class for exponential backoff of operations which
+ * can fail and can be retried.
+ */
+public class ExponentialBackoff {
+    private static final int DEFAULT_MAX_RETRIES = 3;
+    private static final Duration INITIAL_WAIT_PERIOD = Duration.of(250, ChronoUnit.MILLIS);
+
+    private final int maxRetries;
+    private final Duration initialWaitPeriod;
+
+    public ExponentialBackoff() {
+        this(DEFAULT_MAX_RETRIES, INITIAL_WAIT_PERIOD);
+    }
+
+    private ExponentialBackoff(int maxRetries, Duration initialWaitPeriod) {
+        if (maxRetries < 1) {
+            throw new IllegalArgumentException("Max retries must be at least 1");
+        }
+        if (initialWaitPeriod.isNegative() || initialWaitPeriod.isZero()) {
+            throw new IllegalArgumentException("Initial backoff wait delay must be strictly positive");
+        }
+        this.maxRetries = maxRetries;
+        this.initialWaitPeriod = initialWaitPeriod;
+    }
+
+    public static ExponentialBackoff get() {
+        return new ExponentialBackoff();
+    }
+
+    /**
+     * The maximum number of retries.
+     *
+     * @return an exponential backoff with the specified number of retries
+     */
+    public ExponentialBackoff withMaxRetries(int maxRetries) {
+        return new ExponentialBackoff(maxRetries, initialWaitPeriod);
+    }
+
+    /**
+     * The initial backoff duration, that is to say the time we will wait
+     * before the first retry (there's no wait for the initial attempt).
+     *
+     * @param duration the duration for the first retry
+     * @return an exponential backoff with the specified initial wait period
+     */
+    public ExponentialBackoff withInitialWaitPeriod(Duration duration) {
+        return new ExponentialBackoff(maxRetries, duration);
+    }
+
+    /**
+     * Executes an operation which returns a result. Retries a maximum number of
+     * times by multiplying the delay between each attempt by 2.
+     * @param supplier the operation to execute
+     * @return the result of the operation
+     * @param <T> the type of the result
+     */
+    public <T> T supply(FailableSupplier<T> supplier) {
+        int attempts = maxRetries + 1;
+        Duration waitPeriod = initialWaitPeriod;
+        Exception last = null;
+        while (attempts > 0) {
+            try {
+                return supplier.get();
+            } catch (Exception ex) {
+                last = ex;
+                attempts--;
+                try {
+                    Thread.sleep(waitPeriod.toMillis());
+                    waitPeriod = waitPeriod.multipliedBy(2);
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    throw new RetriableOperationFailedException("Thread was interrupted", e);
+                }
+            }
+        }
+        throw new RetriableOperationFailedException("Operation failed after " + maxRetries + " retries", last);
+    }
+
+    /**
+     * Executes an operation which doesn't return any result, until it passes,
+     * with this exponential backoff parameters.
+     * See {@link #supply(FailableSupplier)} for an operation which returns a result.
+     * @param operation the operation to execute.
+     */
+    public void execute(FailableOperation operation) {
+        supply(() -> {
+            operation.run();
+            return null;
+        });
+    }
+
+    @FunctionalInterface
+    public interface FailableOperation {
+        void run() throws Exception;
+    }
+
+    @FunctionalInterface
+    public interface FailableSupplier<T> {
+        T get() throws Exception;
+    }
+
+    public static final class RetriableOperationFailedException extends RuntimeException {
+                public RetriableOperationFailedException(String message, Throwable cause) {
+            super(message, cause);
+        }
+    }
+}

--- a/common/utils/src/main/java/org/graalvm/buildtools/utils/ExponentialBackoff.java
+++ b/common/utils/src/main/java/org/graalvm/buildtools/utils/ExponentialBackoff.java
@@ -24,13 +24,13 @@ import java.time.temporal.ChronoUnit;
  */
 public class ExponentialBackoff {
     private static final int DEFAULT_MAX_RETRIES = 3;
-    private static final Duration INITIAL_WAIT_PERIOD = Duration.of(250, ChronoUnit.MILLIS);
+    private static final Duration DEFAULT_INITIAL_WAIT_PERIOD = Duration.of(250, ChronoUnit.MILLIS);
 
     private final int maxRetries;
     private final Duration initialWaitPeriod;
 
     public ExponentialBackoff() {
-        this(DEFAULT_MAX_RETRIES, INITIAL_WAIT_PERIOD);
+        this(DEFAULT_MAX_RETRIES, DEFAULT_INITIAL_WAIT_PERIOD);
     }
 
     private ExponentialBackoff(int maxRetries, Duration initialWaitPeriod) {

--- a/common/utils/src/test/java/org/graalvm/buildtools/utils/ExponentialBackoffTest.java
+++ b/common/utils/src/test/java/org/graalvm/buildtools/utils/ExponentialBackoffTest.java
@@ -1,3 +1,18 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 package org.graalvm.buildtools.utils;
 
 import org.junit.jupiter.api.DisplayName;

--- a/common/utils/src/test/java/org/graalvm/buildtools/utils/ExponentialBackoffTest.java
+++ b/common/utils/src/test/java/org/graalvm/buildtools/utils/ExponentialBackoffTest.java
@@ -1,0 +1,69 @@
+package org.graalvm.buildtools.utils;
+
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+import java.time.Duration;
+import java.time.temporal.ChronoUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class ExponentialBackoffTest {
+    @Test
+    @DisplayName("executed a passing operation")
+    void simpleExecution() {
+        AtomicBoolean success = new AtomicBoolean();
+        ExponentialBackoff.get().execute(() -> success.set(true));
+        assertTrue(success.get());
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = {1, 3})
+    @DisplayName("retries expected amount of times")
+    void countRetries(int retries) {
+        AtomicInteger count = new AtomicInteger();
+        assertThrows(ExponentialBackoff.RetriableOperationFailedException.class, () -> ExponentialBackoff.get().withMaxRetries(retries)
+            .execute(() -> {
+                count.incrementAndGet();
+                throw new RuntimeException();
+            }));
+        assertEquals(retries + 1, count.get());
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = {1, 3})
+    @DisplayName("passes after one retry")
+    void passAfterRetry(int retries) {
+        AtomicInteger count = new AtomicInteger();
+        int result = ExponentialBackoff.get().withMaxRetries(retries)
+            .supply(() -> {
+                if (count.getAndIncrement() == 0) {
+                    throw new RuntimeException();
+                }
+                return 200;
+            });
+        assertEquals(2, count.get());
+        assertEquals(200, result);
+    }
+
+    @Test
+    @DisplayName("can configure initial backoff time")
+    void canConfigureInitialBackoffTime() {
+        double sd = System.currentTimeMillis();
+        assertThrows(ExponentialBackoff.RetriableOperationFailedException.class, () -> ExponentialBackoff.get()
+            .withMaxRetries(4)
+            .withInitialWaitPeriod(Duration.of(1, ChronoUnit.MILLIS))
+            .execute(() -> {
+                throw new RuntimeException();
+            }));
+        double duration = System.currentTimeMillis() - sd;
+        assertTrue(duration < 100);
+    }
+
+}

--- a/docs/src/docs/asciidoc/index.adoc
+++ b/docs/src/docs/asciidoc/index.adoc
@@ -19,6 +19,12 @@ If you are using alternative build systems, see <<alternative-build-systems.adoc
 [[changelog]]
 == Changelog
 
+=== Release 0.10.3
+
+==== Gradle plugin
+
+- Add retries when downloading the metadata repository when using a URL directly
+
 === Release 0.10.1
 
 - Mark additional JUnit 5 types for build-time initialization for compatibility with Native Image's `--strict-image-heap` option.

--- a/docs/src/docs/asciidoc/index.adoc
+++ b/docs/src/docs/asciidoc/index.adoc
@@ -25,6 +25,10 @@ If you are using alternative build systems, see <<alternative-build-systems.adoc
 
 - Add retries when downloading the metadata repository when using a URL directly
 
+==== Maven plugin
+
+- Add retries when downloading the metadata repository when using a URL directly
+
 === Release 0.10.1
 
 - Mark additional JUnit 5 types for build-time initialization for compatibility with Native Image's `--strict-image-heap` option.

--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -1,6 +1,6 @@
 [versions]
 # Project versions
-nativeBuildTools = "0.10.2"
+nativeBuildTools = "0.10.3-SNAPSHOT"
 metadataRepository = "0.3.8"
 
 # External dependencies

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/NativeImagePlugin.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/NativeImagePlugin.java
@@ -466,6 +466,12 @@ public class NativeImagePlugin implements Plugin<Project> {
                     spec.getParameters().getUri().set(repositoryExtension.getUri().map(serializableTransformerOf(configuredUri -> computeMetadataRepositoryUri(project, repositoryExtension, m -> logFallbackToDefaultUri(m, logger)))));
                     spec.getParameters().getCacheDir().set(
                             new File(project.getGradle().getGradleUserHomeDir(), "native-build-tools/repositories"));
+                    spec.getParameters().getBackoffMaxRetries().convention(
+                        GradleUtils.intProperty(project.getProviders(), "exponential.backoff.max.retries", 3)
+                    );
+                    spec.getParameters().getInitialBackoffMillis().convention(
+                        GradleUtils.intProperty(project.getProviders(), "exponential.backoff.initial.delay", 100)
+                    );
                 });
     }
 

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/GraalVMReachabilityMetadataService.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/GraalVMReachabilityMetadataService.java
@@ -40,6 +40,7 @@
  */
 package org.graalvm.buildtools.gradle.internal;
 
+import org.graalvm.buildtools.utils.ExponentialBackoff;
 import org.graalvm.buildtools.utils.FileUtils;
 import org.graalvm.reachability.DirectoryConfiguration;
 import org.graalvm.reachability.GraalVMReachabilityMetadataRepository;
@@ -66,6 +67,7 @@ import java.nio.channels.Channels;
 import java.nio.channels.ReadableByteChannel;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.time.Duration;
 import java.util.Collection;
 import java.util.Map;
 import java.util.Objects;
@@ -85,6 +87,10 @@ public abstract class GraalVMReachabilityMetadataService implements BuildService
     protected abstract FileSystemOperations getFileOperations();
 
     public interface Params extends BuildServiceParameters {
+        Property<Integer> getBackoffMaxRetries();
+
+        Property<Integer> getInitialBackoffMillis();
+
         Property<LogLevel> getLogLevel();
 
         Property<URI> getUri();
@@ -124,14 +130,16 @@ public abstract class GraalVMReachabilityMetadataService implements BuildService
                         throw new RuntimeException(e);
                     }
                 }
-
-                try (ReadableByteChannel readableByteChannel = Channels.newChannel(uri.toURL().openStream())) {
-                    try (FileOutputStream fileOutputStream = new FileOutputStream(zipped)) {
-                        fileOutputStream.getChannel().transferFrom(readableByteChannel, 0, Long.MAX_VALUE);
-                    }
-                } catch (IOException e) {
-                    throw new RuntimeException(e);
-                }
+                ExponentialBackoff.get()
+                    .withMaxRetries(getParameters().getBackoffMaxRetries().get())
+                    .withInitialWaitPeriod(Duration.ofMillis(getParameters().getInitialBackoffMillis().get()))
+                    .execute(() -> {
+                        try (ReadableByteChannel readableByteChannel = Channels.newChannel(uri.toURL().openStream())) {
+                            try (FileOutputStream fileOutputStream = new FileOutputStream(zipped)) {
+                                fileOutputStream.getChannel().transferFrom(readableByteChannel, 0, Long.MAX_VALUE);
+                            }
+                        }
+                    });
             }
             return newRepositoryFromZipFile(cacheKey, zipped, logLevel);
         }

--- a/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/GradleUtils.java
+++ b/native-gradle-plugin/src/main/java/org/graalvm/buildtools/gradle/internal/GradleUtils.java
@@ -48,6 +48,8 @@ import org.gradle.api.file.ConfigurableFileCollection;
 import org.gradle.api.file.FileCollection;
 import org.gradle.api.plugins.JavaPlugin;
 import org.gradle.api.plugins.JavaPluginExtension;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.provider.ProviderFactory;
 import org.gradle.api.tasks.SourceSet;
 import org.gradle.api.tasks.SourceSetContainer;
 import org.gradle.util.GradleVersion;
@@ -77,17 +79,28 @@ public class GradleUtils {
         ConfigurableFileCollection transitiveProjectArtifacts = project.getObjects().fileCollection();
         transitiveProjectArtifacts.from(findMainArtifacts(project));
         transitiveProjectArtifacts.from(findConfiguration(project, name)
-                .getIncoming()
-                .artifactView(view -> view.componentFilter(ProjectComponentIdentifier.class::isInstance))
-                .getFiles());
+            .getIncoming()
+            .artifactView(view -> view.componentFilter(ProjectComponentIdentifier.class::isInstance))
+            .getFiles());
         return transitiveProjectArtifacts;
     }
 
     public static FileCollection findMainArtifacts(Project project) {
         return findConfiguration(project, JavaPlugin.RUNTIME_ELEMENTS_CONFIGURATION_NAME)
-                .getOutgoing()
-                .getArtifacts()
-                .getFiles();
+            .getOutgoing()
+            .getArtifacts()
+            .getFiles();
     }
 
+    public static Provider<Integer> intProperty(ProviderFactory providers, String propertyName, int defaultValue) {
+        return stringProperty(providers, propertyName)
+            .map(Integer::parseInt)
+            .orElse(defaultValue);
+    }
+
+    private static Provider<String> stringProperty(ProviderFactory providers, String propertyName) {
+        return providers.systemProperty(propertyName)
+            .orElse(providers.gradleProperty(propertyName))
+            .orElse(providers.environmentVariable(propertyName.replace('.', '_').toUpperCase()));
+    }
 }

--- a/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/AbstractNativeMojo.java
+++ b/native-maven-plugin/src/main/java/org/graalvm/buildtools/maven/AbstractNativeMojo.java
@@ -59,6 +59,7 @@ import org.eclipse.aether.resolution.DependencyResolutionException;
 import org.eclipse.aether.resolution.DependencyResult;
 import org.graalvm.buildtools.VersionInfo;
 import org.graalvm.buildtools.maven.config.MetadataRepositoryConfiguration;
+import org.graalvm.buildtools.utils.ExponentialBackoff;
 import org.graalvm.buildtools.utils.FileUtils;
 import org.graalvm.reachability.DirectoryConfiguration;
 import org.graalvm.reachability.GraalVMReachabilityMetadataRepository;
@@ -73,6 +74,7 @@ import java.net.URISyntaxException;
 import java.net.URL;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.time.Duration;
 import java.util.HashSet;
 import java.util.Optional;
 import java.util.Set;
@@ -103,6 +105,12 @@ public abstract class AbstractNativeMojo extends AbstractMojo {
 
     @Parameter(alias = "metadataRepository")
     protected MetadataRepositoryConfiguration metadataRepositoryConfiguration;
+
+    @Parameter(defaultValue = "3")
+    protected int metadataRepositoryMaxRetries;
+
+    @Parameter(defaultValue = "100")
+    protected int metadataRepositoryInitialBackoffMillis;
 
     protected final Set<DirectoryConfiguration> metadataRepositoryConfigurations;
 
@@ -207,7 +215,11 @@ public abstract class AbstractNativeMojo extends AbstractMojo {
                 }
             }
 
-            return downloadMetadataRepo(destinationRoot, targetUrl);
+            URL finalTargetUrl = targetUrl;
+            return ExponentialBackoff.get()
+                .withMaxRetries(metadataRepositoryMaxRetries)
+                .withInitialWaitPeriod(Duration.ofMillis(metadataRepositoryInitialBackoffMillis))
+                .supply(() -> downloadMetadataRepo(destinationRoot, finalTargetUrl));
         }
     }
 

